### ISSUE TITLE
Support for gprc asyncIO

### ISSUE THREAD

**Issue by pixelsoccupied** [d5h-foss/grpc-interceptor#8]
Hello, any plans to support asyncIO intercepter? https://grpc.github.io/grpc/python/grpc_asyncio.html#grpc.aio.ServerInterceptor

Best

**Comment by d5h** [comment]
Good idea. Thanks for bringing it up. I'll look into it (not sure on timing though).

**Comment by rokel** [comment]
Thanks for a great library. I had a quick go at just copy-pasting the `ServerInterceptor` and sprinkling in some `async`/`await` and amazingly it seemed to just work, but I've not used raw grpc interceptors before so don't fully understand if there are any other ramifications to doing this.

```python
class ServerInterceptorAsync(grpc.aio.ServerInterceptor, metaclass=abc.ABCMeta):

    @abc.abstractmethod
    async def intercept(
        self,
        method: Callable,
        request: Any,
        context: grpc.aio.ServicerContext,
        method_name: str,
    ) -> Any:

        return await method(request, context)  # pragma: no cover


    async def intercept_service(self, continuation: Callable[[grpc.HandlerCallDetails], Awaitable[grpc.RpcMethodHandler]], handler_call_details):
 
        next_handler = await continuation(handler_call_details)
        handler_factory, next_handler_method = _get_factory_and_method(next_handler)

        async def invoke_intercept_method(request, context):
            method_name = handler_call_details.method
            return await self.intercept(next_handler_method, request, context, method_name,)

        return handler_factory(
            invoke_intercept_method,
            request_deserializer=next_handler.request_deserializer,
            response_serializer=next_handler.response_serializer,
        )
```

Example usage, just add `async def` and `await` the method:
```python
class RequestLoggingInterceptorAsync(ServerInterceptorAsync):
    async def intercept(
        self,
        method: Callable,
        request: Any,
        context: grpc.aio.ServicerContext,
        method_name: str,
    ) -> Any:
        logging.info(f"REQUEST {method_name}")
        response = await method(request, context)
        logging.info(f"RESPONSE {method_name}")

        return response
```

I don't use the clients so don't know if it will be as simple for them.

**Comment by d5h** [comment]
Ok, I started looking into it. @rokel your solution seems OK. However, I'd like to find a way to do it without copying all the code between sync and async versions. I actually don't mind copying the `ServerInterceptor` code since it's pretty short. But copying all the test code feels like too much. I made some progress on a non-copy/paste solution for tests, but it's a little messy and not quite right yet. I'll need another free weekend or so to work out the kinks.

**Comment by BlakeHyde** [comment]
@d5h has there been any progress on this? Glad to help if need be.

**Comment by d5h** [comment]
Hi @BlakeHyde. Yea, I made some progress on it, but got stuck on async streaming functionality. If you'd like to help push it over the finish line, that would be awesome! I pushed an [`async`](https://github.com/d5h-foss/grpc-interceptor/compare/master...async) branch with my work in progress as a starting point.

It adds a few things:
* An [`AsyncServerInterceptor`](https://github.com/d5h-foss/grpc-interceptor/blob/async/src/grpc_interceptor/server.py#L67) for the core functionality.
* An [`AsyncDummyService`](https://github.com/d5h-foss/grpc-interceptor/blob/async/src/grpc_interceptor/testing/dummy_client.py#L75) to use as a test service.
* An [`aio_server`](https://github.com/d5h-foss/grpc-interceptor/blob/async/src/grpc_interceptor/testing/dummy_client.py#L131) flag to `dummy_channel` in order to create the `AsyncDummyService`. This creates an asycio loop and runs the service on it in a [background thread](https://github.com/d5h-foss/grpc-interceptor/blob/async/src/grpc_interceptor/testing/dummy_client.py#L159).
* Async versions of all the test interceptors ([example](https://github.com/d5h-foss/grpc-interceptor/blob/async/tests/test_server.py#L30)).
* Parameterizes each test with an `aio` flag. This is used to decide whether to create a sync or async version of the interceptor, and also gets passed to the `aio_server` flag mentioned above to create a sync or async version of the service ([example](https://github.com/d5h-foss/grpc-interceptor/blob/async/tests/test_server.py#L111-L119)).

I did it this way so that the tests can mostly stay the same. We'll just run a sync version and an async version of each test. This should work in theory because the tests create a client that connects to a server, and then checks responses. The client shouldn't care whether the server is sync or async.

This all currently works for the unary-unary calls. However, three of the streaming calls fail. I tried for a while to figure out what was wrong, but I don't use either asyncio or streaming gRPC at all, so am not very familiar with them.

If you can fix these, it would be much appreciated! Running `make test` (after `poetry install`) will show what's failing.

By the way, this branch only adds server-side async support, not client-side async support.

**Comment by BlakeHyde** [comment]
@d5h Okay, I took a look at it over the weekend and I suspect that, unfortunately, this is not going to be fixable with the current design. The issue is that the differentiation between streaming/unary has more implications for async than non-async code, namely the async generator vs. the normal coroutine. As it stands, I can reduce the failures to 1 by deciding within `intercept_service` whether to yield or return... but that last failure is deep in C code and seems to imply that there needs to be differentiation based on both `handler.request_streaming` and `handler.response_streaming`. 😬 

If you'd like to see, the only change necessary to get to this state is:

```python
        if next_handler.response_streaming or next_handler.request_streaming:
            async def invoke_intercept_method(request, context):
                res = self.intercept(
                    next_handler_method,
                    request,
                    context,
                    method_name,
                )
                async for r in res:
                    yield r
        else:
            async def invoke_intercept_method(request, context):
                return await self.intercept(
                    next_handler_method,
                    request,
                    context,
                    method_name,
                )
```

**Comment by l4d2boomer** [comment]
I am waiting for the update :)

**Comment by d5h** [comment]
I finally got async server interceptors working. But I'm still in the process of testing some cases, documentation, and doing some cleanup. Hopefully I can release it in the next couple of weeks, time permitting. In the meanwhile, if you'd like to beta test it, you can install it with:

```
pip install git+https://github.com/d5h-foss/grpc-interceptor.git@async
```

Async streaming gRPC has some wrinkles, which makes the interceptors a little less smooth than sync. If you're only dealing with unary calls, then you can do the natural:

```python
from grpc_interceptor import AsyncServerInterceptor

class AsyncInterceptor(AsyncServerInterceptor):
    # works for unary
    async def intercept(self, method, request, context, method_name):
        response = await method(request, context)
        # do something
        return response
```

However, if you have any streaming methods, you cannot `await` them. Assuming you want to do something with the response, you can make your interceptor an async generator:

```python
class AsyncInterceptor(AsyncServerInterceptor):
    # works for streaming
    async def intercept(self, method, request, context, method_name):
        async for response in method(request, context):
            # do something
            yield response
```

To support both unary and streaming, you'll have to do something like this:

```python
class AsyncInterceptor(AsyncServerInterceptor):
    # works for both unary and streaming

    async def _handle_streaming(self, iterator):
        async for resp in iterator:
            # do something with response
            yield resp

    async def intercept(self, method, request, context, method_name):
        response_or_iterator = method(request, context)
        if hasattr(response_or_iterator, "__aiter__"):
            return self._handle_streaming(response_or_iterator)  # no await
        else:
            response = await response_or_iterator
            # do something with response
            return response_or_iterator
```

This is because streaming endpoints are typically implemented with `async def + yield`. When you call such a method, you get back an `async_generator`. This is not `await`-able (though you can `async for` loop over it). A unary endpoint is implemented with `async def + return`, which results in a `coroutine` when called, which you _can_ `await`. However, it's a syntax error to mix `async def + return` with `async def + yield`. Hence, you need to split the unary case and streaming case into two different methods.

Note that if all your endpoints are unary, but you're using `grpcio-reflection`, you'll still need to support streaming. That's because the reflection API uses it. In this case you could just pass it through and ignore it:

```python
        response_or_iterator = method(request, context)
        if hasattr(response_or_iterator, "__aiter__"):
            return response_or_iterator
        else:
            return await response_or_iterator
```

Also, note that async gRPC added a new `read()` / `write()` [alternative API for streaming](https://github.com/lidizheng/proposal/blob/grpc-python-async-api/L58-python-async-api.md#new-streaming-api---reader--writer-api). I've only tested the `write()` portion of this so far. I'll add a `read()` test when I have more time, and hopefully that works out OK. 🤞 

So to recap, progress has been made, but still a little to do. Feel free to test it and provide feedback.

**Comment by d5h** [comment]
Ok, async server interceptor support has been added in 0.15.0. See the [release notes](https://github.com/d5h-foss/grpc-interceptor/releases/tag/v0.15.0) and [docs](https://grpc-interceptor.readthedocs.io/en/latest/#async-server-interceptors) for more details. Please let me know if you find any issues.
